### PR TITLE
Feature/#88 parsing courses elearning

### PR DESCRIPTION
**Pull request by davidm1766**
Parsing courses from elearning with tutorial how to use it in unit test
Fix mailgun configuration

### CODE REVIEW

File: Backend/Converter/ConverterApiToDomain.cs [hunk 1]
```diff
@@ -55,7 +55,7 @@ private static async Task<Timetable> ConvertTimetableAsync(ScheduleWeekContent s
                             };
                             if (!isTimetableForCourse)
                             {
-                                bl.CourseId = await courseServ.GetOrAddNotExistsCourseId(block.CourseName, courseServ, bl);
+                                bl.CourseId = await courseServ.GetOrAddNotExistsCourseId(block.CourseName, bl);
```
Comment (marektavac) [inline Backend/Converter/ConverterApiToDomain.cs]: Is course name unique?

File: CoursesParser/BaseParser.cs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+
+        public HashSet<string> ParseFaculties()
+        {
+            
+            foreach (var facultyOption in _selectFaculties.ChildNodes)
+            {
+                _facId = facultyOption.Attributes["value"].Value;
+                DownloadAndSaveCourses(1);
```
Comment (marektavac) [inline CoursesParser/BaseParser.cs]: What does it mean 1,2, 3... May be enum would be more clear...

File: Backend/Converter/ConverterApiToDomain.cs (hunk 1 above)
Comment (davidm1766) [reply Backend/Converter/ConverterApiToDomain.cs]: yes, parser remove all duplicite

File: CoursesParser/BaseParser.cs (hunk 2 above)
Comment (davidm1766) [reply CoursesParser/BaseParser.cs]: Is is simulating magic javascript function on the website vzdelavanie - I replace it with Enum

**Comment by davidm1766** [comment]
I think that writing unit test is useless and very hard to write(because of style how magically vzdelavanie works). Vzdelavanie page is complicated - whole system of displaying names of courses is not intuitive and unit test from my opinion dont help us when they change website because it will be necessary to study whole system again.

File: Entities/BaseParser.cs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        private string DownloadJson(string f,string t,string m,string r,
+                                             string o,string z,string c)
+        {
+            string retJson = "";
+            using (WebClient wc = new WebClient())
+            {
+                wc.Encoding = _encoding;
+                retJson = wc.DownloadString($"http://vzdelavanie.uniza.sk/vzdelavanie/plany.php?f={f}&t={t}&m={m}&r={r}&o={o}&z={z}&c={c}");
```
Comment (ZVARKO) [inline Entities/BaseParser.cs]: you have same url on line 32...

**Comment by davidm1766** [comment]
@ZVARKO  please check it - @dobi28  should implement email templates

File: CoursesParser/BaseParser.cs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+using System.Diagnostics;
+using System.Net;
+using System.Text;
+
+
+namespace CoursesParser
+{
+    public class BaseParser
+    {
+        private string _url = "http://vzdelavanie.uniza.sk/vzdelavanie/plany.php";
```
Comment (ZVARKO) [inline CoursesParser/BaseParser.cs]: use Pascal case for constants

File: Tests/BackendTest/CourseParserTest.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    public class CourseParserTest : IClassFixture<Mongo2GoFixture>
+    {
+        private readonly Mongo2GoFixture _mongoFixture;
+        public CourseParserTest(Mongo2GoFixture mongoFixture)
+        {
+            _mongoFixture = mongoFixture;
+        }
+
+        [Fact]
+        public async Task ParseFacultiesAsync()
```
Comment (ZVARKO) [inline Tests/BackendTest/CourseParserTest.cs]: Why is this test commented out?

File: WebAPI/appsettings.json [hunk 6]
```diff
@@ -20,15 +20,15 @@
...
-    "RequireDigit": "true",
-    "RequiredLength": "8",
-    "RequireNonAlphanumeric": "false",
-    "RequireUppercase": "true",
-    "RequireLowercase": "true",
-    "RequireConfirmedEmail": "true",
-    "DefaultLockoutTimeSpan": "30",
-    "MaxFailedAccessAttempts": "10",
-    "RequireUniqueEmail": "true" 
+    "IdentitySettings": {
```
Comment (ZVARKO) [inline WebAPI/appsettings.json]: undo this change

File: Tests/IntegrationTest/TestFixture.cs [hunk 7]
```diff
@@ -87,8 +88,8 @@ private void MockServices(IServiceCollection services)
                 services.Remove(serviceDescriptor);
             }
             var emailServiceMock = new Mock<IEmailService>();
-            emailServiceMock.Setup(x => x.SendRegistrationConfirmationEmail(It.IsAny<string>(), It.IsAny<string>()))
-                           .Verifiable();
+            var email = new Email() { Body= It.IsAny<string>(),Subject=It.IsAny<string>(),ToEmail=It.IsAny<string>()};
```
Comment (ZVARKO) [inline Tests/IntegrationTest/TestFixture.cs]: use ctrl+kd to format code

File: Backend/EmailService.cs [hunk 8]
```diff
@@ -19,19 +20,32 @@ public EmailService(IOptions<MailingSettings> emailSettings)
...
-                string body = $"Pre potvrdenie účtu klikni na tento <a href='{confirmationLink}'>odkaz</a>.";
-                MailMessage message = new MailMessage(_emailSettings.Username, receiverEmail, "Swapify - potvrdenie registrácie", body);
+                client.EnableSsl = true;                
+                MailMessage message = new MailMessage(_emailSettings.Username, email.ToEmail, email.Subject, email.Body);
                 message.IsBodyHtml = true;
                 client.Send(message);
             }
         }
+
+        //public virtual void SendRegistrationConfirmationEmail(string receiverEmail, string confirmationLink)
```
Comment (ZVARKO) [inline Backend/EmailService.cs]: Why is this code commented out? Remove it.

File: Tests/BackendTest/CourseParserTest.cs (hunk 5 above)
Comment (davidm1766) [reply Tests/BackendTest/CourseParserTest.cs]: OK

File: CoursesParser/BaseParser.cs (hunk 4 above)
Comment (ZVARKO) [reply CoursesParser/BaseParser.cs]: still not resolved

File: WebAPI/appsettings.json (hunk 6 above)
Comment (ZVARKO) [reply WebAPI/appsettings.json]: still not resolved

Comment (davidm1766) [reply WebAPI/appsettings.json]: image
it is just bad compare I remove only MailGuns section

File: WebAPI/appsettings.json [hunk 9]
```diff
@@ -12,6 +12,9 @@
       }
     }
   },
+  "PathSettings": {
+    "CoursesJsonPath": "..\\CoursesParser\\ActualCoursesJson\\courses_2018_11_20.json"
```
Comment (ZVARKO) [inline WebAPI/appsettings.json]: do we really need this in settings file? Path is wrong. You should copy the file to the output directory and load it from there.

File: Backend/Settings/PathSettings.cs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+        public PathSettings()
+        {
+
+        }
+
+        public override void Validate()
+        {
+            if (string.IsNullOrEmpty(CoursesJsonPath))
+                Errors.AppendLine($"Setting {nameof(CoursesJsonPath)} is missing in {nameof(PathSettings)} configuration section.");
```
Comment (ZVARKO) [inline Backend/Settings/PathSettings.cs]: did you test it before push? This will not work because you forgot to call CheckErrors() method and you also forgot to register PathSettings in startup file... so this code will never execute.

File: Backend/DbSeed/DbSeed.cs [hunk 11]
```diff
@@ -36,5 +41,28 @@ public static void CreateTestingUser(IServiceProvider serviceProvider)
...
             }
         }
+
+        public static void CreateTestingCourses(IServiceProvider serviceProvider)
+        {
+            var dbService = serviceProvider.GetRequiredService<IMongoDatabase>();
+            var courseCollection = dbService.GetCollection<Course>(nameof(Course));
+            
+            var path = serviceProvider.GetRequiredService<IOptions<PathSettings>>();
+            var json = File.ReadAllText(path.Value.CoursesJsonPath);
```
Comment (ZVARKO) [inline Backend/DbSeed/DbSeed.cs]: File.ReadAllText is throwing a lot of exceptions... Check if file exists before calling ReadAllText().

File: CoursesParser/BaseParser.cs [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+using System.Diagnostics;
+using System.Net;
+using System.Text;
+
+namespace CoursesParser
+{
+    public class BaseParser
+    {
+        private const string Url = "http://vzdelavanie.uniza.sk/vzdelavanie/plany.php";
+        private HtmlDocument _document;
```
Comment (ZVARKO) [inline CoursesParser/BaseParser.cs]: why did you remove readonly kewords from here? Why is _document declared as field?

File: Backend/DbSeed/DbSeed.cs (hunk 11 above)
Comment (davidm1766) [reply Backend/DbSeed/DbSeed.cs]: this function is in try catch block in startup - it is useless handling it on two places

File: WebAPI/appsettings.json (hunk 9 above)
Comment (davidm1766) [reply WebAPI/appsettings.json]: Problem is that there are two different project = two different outputs, if I would like to copy file from one project output to another command I need to use post build event and command for copy will be depend on operating system
